README
NAME
    SenseClusters - Cluster similar contexts using co-occurrence matrices
    and Latent Semantic Analysis

SYNOPSIS
    SenseClusters is a suite of Perl programs that supports unsupervised
    clustering of similar contexts. It relies on it's own native
    methodology, and also provides support for Latent Semantic Analysis.

    SenseClusters is a complete system that takes users from preprocessing
    of raw text to providing clustered output. It supports the selection of
    features, the creation of various kinds of context representations,
    dimensionality reduction by singular value decomposition, clustering,
    and analysis of results.

    SenseClusters integrates specialized tools such as the Ngram Statistics
    Package (Text::NSP), SVDPACK, the Perl Data Language (PDL) and CLUTO to
    provide a variety of choices and high efficiency at each step in its
    processing.

OVERVIEW
    SenseClusters supports several different methods of clustering contexts.
    These include the native SenseClusters methodology, which is based on
    the use of first and second order representations of contexts. It also
    includes support for clustering lexical features using the native
    SenseClusters methodology or Latent Semantic Analysis.

    SenseClusters is based strictly on lexical features and does not rely on
    any manually created training data or external knowledge sources, and as
    such is language independent. The only requirement is that the language
    should be able to be tokenized via Perl regular expressions, which can
    be specified by the user. In fact, tokenization is so flexible that
    features could consist of characters, pairs of characters, etc.

    SenseClusters can be applied to the problem of discriminating word
    meanings or ambiguous names, using the target or head word
    representation. This is sometimes also called "headed" data, where each
    context is centered around the given target whose meanings are to be
    discovered. In this case the contexts that contain the given target word
    are clustered, and each cluster is assumed to correspond to a different
    meaning of that word.

    SenseClusters can also be applied to the problem of grouping short units
    of text that have no target or head (which is sometimes referred to as a
    "headless" representation. In this case there is no head or center to
    the context, so the entire context is being clustered to determine the
    meaning or topic of the context as a whole. Email categorization or news
    article clustering are examples of problems that could be approached
    using headless data.

    SenseClusters will automatically determine the number of clusters in the
    data based on a number of different automatic stopping measures we have
    developed, three of which are based on clustering criterion function,
    and one which is an adaptation of the well-known Gap Statistic.

    SenseClusters can also be applied to the problem of clustering words or
    lexical features, in hopes of discovering synonyms, antonyms, or other
    classes of words.

    Broadly speaking, SenseClusters can be used for any task that requires
    the recognition of contextually similar units of text, or words that
    occur in similar contexts.

DOCUMENTATION
    All programs have inline source code documentation written in pod style
    and this can be browsed from command line as a man page or using the
    'perldoc' command. For example, 'man bitsimat.pl' or 'perldoc
    bitsimat.pl' will displayed the documentation for the bitsimat.pl
    program. Each program also has a --help option to provide information
    about program options.

    You can see all of the modules and their associated documentation at
    README.Toolkit.

GETTING STARTED
    You might first like to run the demonstration scripts in samples/
    directory to get an idea of SenseClusters' usage and functionality, or
    try the web interface that is provided at
    <http://senseclusters.sourceforge.net>.

    samples/ contains scripts that utilize the wrapper program
    discriminate.pl that calls various other programs from the package to
    run a complete experiment. It also contains examples where specialized
    experiments are constructed directly from the programs provided in the
    package. In general it would be useful to consult the flowcharts in
    doc/Flowcharts to understand the overall structure of the package.

    The web interface provides an intuitive means of formulating and running
    discriminate.pl commands, so the use of the web interface and certainly
    be instructive in terms of how to formulate discriminate.pl commands.

    The contexts that you wish to cluster must be in Senseval-2 format. This
    is a simple XML markup that indicates the beginning and end of each
    context, and allows you to specify a target word and a "correct"
    categorization of the context, if you know that information. There is a
    pre-processing program text2sval.pl in Toolkit/preprocess/plain/ that
    converts plain text data (with a single context on each line) into
    Senseval-2 format. There is also a large amount of sample data that is
    already in Senseval-2 format available at
    <http://senseclusters.sourceforge.net>

    You can also (optionally) provide a separate training file in plain text
    format to be used as the feature selection data. If you don't do this,
    then the features will be selected from the contexts to be clustered.

PACKAGE ORGANIZATION
    After downloading and unpacking SenseClusters, you should find following
    files/directories within SenseClusters' directory.

    README, INSTALL, CHANGES, TODO, FAQ
        Read-only copies of documentation found in doc/*.pod

    GPL.txt
        A copy of the GNU General Public License, the terms under which
        SenseClusters is distributed.

    FDL.txt
        A copy of the GNU Free Documentation License, the terms under which
        the documentation of SenseClusters is distributed.

    discriminate.pl
        A wrapper program that acts as a driver for many other programs in
        the package. It clusters the given text instances based on their
        contextual similarities.

    Makefile.PL
        Generates a Makefile on running 'perl Makefile.PL'.

    doc/
        Contains various *.pod files that are kept in a read only form in
        the top level directory.

        doc/Flowcharts/ contains flow diagrams that illustrate how to put
        together the programs provided in SenseClusters' Toolkit with other
        packages like NSP, SVDPACK and CLUTO to run experiments without
        wrappers.

    lib/
        A stub for the Text::Similarity perl module. At present
        SenseClusters is oriented about the command line, so this is mostly
        for the benefit of CPAN indexing.

    * Testing/
        A directory of test cases written as C-shell scripts that will test
        if the package is installed properly or not.

    t/  A stub directory created by h2xs - future site of test cases rather
        than /Testing

    samples/
        A directory of scripts that demonstrate SenseClusters' usage and
        functionality.

    External/
        Contains a modified version of SVDPACKC, and a script that can be
        run to automatically install it, and retrieve and install Cluto.

    Toolkit/
        A directory of Perl programs implemented and used by SenseClusters.
        Users who are interested to use SenseClusters' tools individually
        and separately without using the wrapper programs are encouraged to
        browse through the Toolkit and Toolkit.pod.

    * Web/
        Contains an easy to use and install web interface for SenseClusters.

CONTACT US
    Please join our mailing lists to participate in the package related
    discussions, to post your questions or bugs and also to suggest
    enhancements to the package functionality.

    To subscribe to the user's mailing list, visit :
    <http://lists.sourceforge.net/lists/listinfo/senseclusters-users>

    To subscribe to a low volume news mailing list, visit :
    <http://lists.sourceforge.net/lists/listinfo/senseclusters-news>

    To subscribe to the developer's mailing list, visit :
    <http://lists.sourceforge.net/lists/listinfo/senseclusters-developers>

SEE ALSO
    <http://senseclusters.sourceforge.net/>

AUTHORS
     Ted Pedersen
     University of Minnesota, Duluth
     tpederse at d.umn.edu
     http://www.d.umn.edu/~tpederse/

     Amruta Purandare
     University of Pittsburgh

     Anagha Kulkarni
     Carnegie-Mellon University

     Mahesh Joshi
     Carnegie-Mellon University

ACKNOWLEDGMENTS
    This work has been partially supported by a National Science Foundation
    Faculty Early CAREER Development award (Grant #0092784).

COPYRIGHT
    Copyright (c) 2003-2008, Ted Pedersen, Amruta Purandare, Anagha
    Kulkarni, and Mahesh Joshi

    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 2 of the License, or (at your
    option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to

     The Free Software Foundation, Inc.,
     59 Temple Place - Suite 330,
     Boston, MA  02111-1307, USA.



CHANGES
NAME
    CHANGES - Revision history of SenseClusters

DESCRIPTION
    1.01
    Released April 6, 2008 (All changes by TDP)

    *   Relaxed test conditions for svdpackout and install.sh to simply
        check if SVDPACKC is producing results.

    *   Removed diagnostic warning from mat2harbo.pl that gave guidelines on
        setting las2.h parameters. A version of las2.h is now being provided
        in /External with fairly large values pre-set, thus making these
        warnings somewhat less necessary.

    *   Changed the interpretation of the --negatives option below to mean
        that if this option is on, then any negative values in svdpackout.pl
        output will be set to zero. By default then, square root values are
        not taken, and negatives are not smoothed.

    *   Introduced --sqrt option for svdpackout.pl to request that values in
        S matrix have their square root taken before recombination. This was
        previously the default, but will not be any longer.

    *   Updated svdpackout.pl to include the following changes, provided by
        Richard Wicentowski in November 2006 (via senseclusters-developers
        list)

        * Added feature to output the component U, S, and V' matrices.
        * Added a new command-line option "--output" with three options:

          * reconstruct - reconstructs the rank-k matrix (default)
          * rowonly - same as --rowonly
          * components - output U, S, V' matrices to U.txt, S.txt, VT.txt

        * Added a new command-line option "--negatives": Allows negative
        values; otherwise all negative values are set to 0 (except in
        component output).
        * New options maintain backward compatibility
        * Updated the documentation.
        * Passes all tests (testA1-A4,B1-B2)

    1.00
    Released March 29, 2008 (All changes by TDP) This is a development
    release, meaning that "regular" users should still use 0.95, our last
    stable release

    *   Updated SYNOPSIS examples in many of the Toolkit programs to make
        their stand alone use more clear

    *   Removed dependency on Math::BigInt, Math::BigFloat, as that has been
        added to Algorithm::RandomMatrixGeneration as of 0.06 (which is
        required).

    *   Raised Text-NSP version requirement to 1.09, since there are a few
        bugs in the test cases for 1.07 and below for 64 bit architectures.

    *   Changed CPAN module prereqs to require modules that only require
        Perl 5.6 or better. This relaxes SenseClusters from requiring Perl
        5.8.5 to 5.6.

    0.98
    Released March 25, 2008 (All changes by TDP) This is a development
    release, meaning that "regular" users should still use 0.95, our last
    stable release

    *   Inserted a /t directory for 'make test'. This is just a stub to be
        built upon.

    *   Modified README.Toolkit to (hopefully) include links to source code
        documentation while displaying all the programs in a single page.
        This is intended to be our replacement for the HTML pages we no
        longer generate.

    *   Modified Makefile.PL to create man pages only of core SenseClusters
        programs. Reduced the number of programs put into /bin as well, only
        including core SenseClusters programs and those needed for testing.

    *   Modified /Testing script completely. Removed alltests.sh subtests.sh
        and the many normal-op.sh and error-op.sh scripts and replaced with
        a single script called ALL-TESTS.sh (using same name as NSP).
        Historically we've had some trouble with the test scripts not
        running due to file permissions and path issues, so I hope this
        fixes that once and for all. The test*.sh files in the
        subdirectories of /Testing are called directly via ALL-TESTS.sh
        using 'csh ./testA*.sh' after the script has changed to the current
        directory of the test script. This does not require that the scripts
        be executable and path issues are minimized. This also allows us to
        not install altests.sh and subtests.sh in /bin (which was our
        solution to path problems, but that seemed a bit extreme).

    *   Added copyright information to .pl files in Web directory, also
        minimal perldoc.

    *   Added short labels to each of the programs in Toolkit, so that these
        descriptions appear in CPAN listing

    *   Removed HTML directory and support from /Docs, since we'll get
        nicely formatted documentation from CPAN. Removed /Docs, which has
        been replaced with /doc

    *   Removed samples from /Docs for regex, stoplist, and token files,
        since those also exist in Demos

    *   Added a directory /External that includes a modified form of the
        SVDPACKC program las2 that we use for SVD. This can be installed via
        a script External/install.sh that will also attempt to retrieve and
        install the Cluto programs scluster and vcluster.

    *   Moved Docs/Flows to doc/Flowcharts, depricated flowchart.fig and
        .pdf in favor of more recent versions

    *   Moved INSTALL from top level to /doc and converted into pod.

    *   Created doc directory which will be home of pod that is stored in
        top level in plain text form. /doc/update-pod.sh will be used to
        convert pod to plain text which should be considered read only.

    *   Added stubs to /lib and /t to provide basic modular framework should
        we decided to develop modules in future

    *   Moved current contents of Changes to /doc/Changelogs and then began
        a new file CHANGES.pod that will be used in future

    0.96
    Released March 22, 2008 (All modifications by TDP) This is a development
    release, meaning that "regular" users should still use 0.95, our last
    stable release

    *   version 0.95 uploaded to CPAN to see how indexing looked. not

Toolkit/evaluate/label.pl
#!/usr/local/bin/perl -w

=head1 NAME

label.pl Assign labels to clusters to maximize agreement 

=head1 SYNOPSIS

Labels the discovered clusters with sense tags such that maximum number of 
contexts are correctly assigned.

=head1 USAGE

label.pl [OPTIONS] PRELABEL

=head1 INPUT

=head2 Required Arguments:

PRELABEL

Should be the output of cluto2label.pl.

Sample CLUTO2LABEL format 

 2
 //	cord  phone   text   div
 C0:	 4       3       0       0
 C1:	 2       2       2       2
 C2:	 1       3       3       2

where the 1st line shows the number of unclustereted instances = 2 

2nd line shows a space separated list of sense classes starting with // mark.

Each line thereafter shows the sense distribution of the instances belonging
to each discovered cluster in the form of a cluster by sense distribution
matrix. A cell value at (i,j) in the matrix shows the number of instances
belonging to cluster Ci that have the sense tag Sj.

Note that each row begins with the cluster id that precedes a colon (:).
Also, the number of sense classes on 2nd line should be same as the number 
of columns in the cluster by sense distribution table.

=head2 Optional Arguments:

=head3 --help

Displays this message.

=head3 --version

Displays the version information.

=head1 OUTPUT

Output shows the sense labels attached to each of the discovered 
clusters along with the score. Score tells the percentage of the total 
number of instances correctly clustered if the clusters are tagged with 
the sense labels as suggested.

Example :

Prelabel file =>

 0
 //      cord    divi    form    phon    prod    text
 C0:     35      26      44      18      23      43
 C1:     64      34      50      43      57      52
 C2:     0       3       1       2       0       3
 C3:     0       0       2       31      0       0
 C4:     1       28      0       4       6       0
 C5:     0       9       3       2       14      2

Label Output =>

 ClusterID -> SenseID
 C0 -> form
 C1 -> cord
 C2 -> text
 C3 -> phon
 C4 -> divi
 C5 -> prod
 Score = 30.67

shows that 

 cluster C0 represents the 'form' sense
 cluster C1 represents the 'cord' sense
 cluster C2 represents the 'text' sense
 cluster C3 represents the 'phon' sense
 cluster C4 represents the 'divi' sense
 and cluster C5 represents the 'prod' sense

Also, 30.67 % of the total instances are in their right sense classes
if the clusters are tagged with this labeling scheme.

=cut

#			===============================
#			COMMAND LINE OPTIONS AND USAGE 
#			=============================== 	

use Algorithm::Munkres;

use Getopt::Long;
GetOptions ("help","version");

#command option for help
if(defined $opt_help)
{
        $opt_help=1;
        &showhelp();
        exit;
}
#version information
if(defined $opt_version)
{
        $opt_version=1;
        &showversion();
        exit;
}

#show minimal usage
if($#ARGV<0)
{
        &minimal();
        exit;
}

#truncate $0 which contains the complete path to
#the program. Keep just the program name
#this is used in the error messages
$0=~s/.*\/(.+)/$1/;

# input file
if(defined $ARGV[0])
{
	$infile=$ARGV[0];
	if(-e $infile)
	{
		open(IN,$infile) || die "Error($0): Error(code=$!) in opening file <$infile>.\n";
	}
	else
	{
		print STDERR "ERROR($0): PRELABEL file <$infile> doesn't exist...\n";
		exit;
	}
}
else
{
	print STDERR "ERROR($0): Please specify the PRELABEL file name...\n";
	exit;
}

# first line of prelabel file should show number of instances thrown or 
# unclustered 
# label.pl should just pass this information to next program that computes 
# precision and recall
$thrown=<IN>;
if(!defined $thrown)
{
	print STDERR "ERROR($0):
        1st line in the PRELABEL file <$infile> should show 
	the number of instances unclustered or 0.\n";
        exit;
}
chomp $thrown;
if(!($thrown=~/^\s*\d+\s*$/))
{
        print STDERR "ERROR($0):
        1st line in the PRELABEL file <$infile> should show 
	the number of instances unclustered or 0.\n";
        exit;
}

# 2nd line of prelabel output should list all the Sense Classes
# senses should be space separated 
$sense_string=<IN>;
if(defined $sense_string && $sense_string=~/\/\//)
{
	$sense_string=$';
	$sense_string=~s/\s+$//g;
        $sense_string=~s/^\s+//g;
        $sense_string=~s/\s+/ /g;
	# stores all sense classes listed on this line 
	@all_senses=split(/\s+/,$sense_string);
}
else
{
	print STDERR "ERROR($0):
	2nd line in the input file <$infile> should list 
	the sense labels starting with //.\n";
	exit;
}

# accept matrix entries row wise
$i=0;
# total is the total #instances
$total=0;
# read the input file with each row on each line
$line_num=0;
while(<IN>)
{
	chomp;
	s/\s+$//g;
	s/^\s+//g;
	s/\s+/ /g;
	if(/^\s*$/)
	{
		next;
	}

	# we use cluster ids only while printing
	# the final output 
	# during processing, we use the serial 
	# numbers 0,1,... for clusters in the order
	# as they appear in the prelabel file
	($cid,$row)=split(/\s*:\s*/);
	push @cluster_ids,$cid;
	
	#extract the matrix cells
	@row_elements=split(/\s+/,$row);

	if($#row_elements!=$#all_senses)
	{
		print STDERR "ERROR($0):
	        Number of columns (". scalar(@row_elements) . ") at line <$line_num> 
	        in PRELABEL file <$infile> doesn't match the number of senses (" . 
             	scalar(@all_senses) . ") specified on Line 2 of the same file.\n";
		exit;
	}
	
	for($cnt = 0; $cnt <= $#row_elements; $cnt++)
	{
	    if($row_elements[$cnt]!~/^[0-9]+$/)
	    {
		print STDERR "ERROR($0): Line <" . $line_num+1 . "> in PRELABEL file <$infile> contains a non-integer matrix value.\n";
		exit;
	    }
	    else
	    {
		$inp_mat[$line_num][$cnt] = -1 * $row_elements[$cnt];
		$total += $row_elements[$cnt];
	    }
	}

	$line_num++;
}

my @soln_mat = ();

assign(\@inp_mat, \@soln_mat);
$clus_total = 0;

print "ClusterID -> SenseID\n";
for($i=0;$i<=$#inp_mat;$i++)
{
    if(defined $inp_mat[$i][$soln_mat[$i]])
    {
	$clus_total += $inp_mat[$i][$soln_mat[$i]];
    }
    if(defined $all_senses[$soln_mat[$i]])
    {
	print $cluster_ids[$i] . " -> " . $all_senses[$soln_mat[$i]] . "\n";
    }

}
if($total != 0)
{
    $score = $clus_total/$total * -100;
    $score = sprintf("%.2f",$score);
    print "Score = $score\n";
}
else
{
    print "Score = NA\n";    

}


#show minimal usage message
sub minimal()
{
        print "Usage: label.pl [OPTIONS] PRELABEL";
        print "\nTYPE label.pl --help for help\n";
}

#show help
sub showhelp()
{
	print "Usage: label.pl [OPTIONS] PRELABEL\n";
	print "Labels the discovered clusters with sense tags such\n";
	print "that maximum number of contexts are correctly assigned.\n";

	print "\nPRELABEL\n";
        print "Should be an output created by cluto2label.pl\n";
        print "showing a cluster by sense distribution matrix.\n\n";

        print "OPTIONS:\n";

        print "--help
	Displays this message.\n";

        print "--version
	Displays the version information.\n";
}

#version information
sub showversion()
{
 	print '$Id: label.pl,v 1.14 2008/03/24 00:56:58 tpederse Exp $'; 
       print "\nCopyright (C) 2002-2006, Ted Pedersen, Amruta Purandare, & Anagha Kulkarni\n";
#        print "label.pl      -       Version 0.11\n";
#	print "Labels the discovered clusters with sense tags.";
#	print "Date of Last Update:	11/30/2004\n";
	
}

=head1 AUTHOR

Ted Pedersen, University of Minnesota, Duluth

Amruta Purandare, University of Pittsburgh

Anagha Kukarni, University of Minnesota, Duluth

=head1 COPYRIGHT

Copyright (c) 2002-2006,

 Anagha Kulkarni, University of Minnesota, Duluth
 kulka020 at d.umn.edu

 Ted Pedersen, University of Minnesota, Duluth
 tpederse at d.umn.edu

 Amruta Purandare, University of Pittsburgh
 amruta at cs.pitt.edu

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to

 The Free Software Foundation, Inc.,
 59 Temple Place - Suite 330,
 Boston, MA  02111-1307, USA.

